File: backend/app/tools/budget.py
from fastapi import APIRouter
from pydantic import BaseModel, Field

tool_name = "budget"
router = APIRouter(prefix=f"/{tool_name}", tags=[tool_name])


class BudgetEstimateRequest(BaseModel):
    guest_count: int = Field(default=8, ge=1)
    per_guest: float = Field(default=14, ge=0)
    categories: list[str] = Field(default_factory=lambda: ["drinks", "snacks", "supplies"])
    currency: str = "EUR"


class BudgetLineItem(BaseModel):
    category: str
    estimate: float


class BudgetEstimateResponse(BaseModel):
    guest_count: int
    per_guest: float
    currency: str
    line_items: list[BudgetLineItem]
    total: float


capabilities = [
    {
        "id": "budget.estimate",
        "tool": "budget",
        "name": "Estimate Budget",
        "description": "Estimate cost from guest count, per-guest spend, and categories.",
        "method": "POST",
        "endpoint": "/budget/estimate",
        "input_model": "BudgetEstimateRequest",
        "output_model": "BudgetEstimateResponse",
        "tags": ["budget", "planning"],
    }
]


@router.post("/estimate", response_model=BudgetEstimateResponse)
def estimate_budget(payload: BudgetEstimateRequest):
    line_items = []

    for category in payload.categories:
        multiplier = 0.25 if category == "decor" else 1
        estimate = round(payload.guest_count * payload.per_guest * multiplier, 2)
        line_items.append(BudgetLineItem(category=category, estimate=estimate))

    total = round(sum(item.estimate for item in line_items), 2)

    return BudgetEstimateResponse(
        guest_count=payload.guest_count,
        per_guest=payload.per_guest,
        currency=payload.currency,
        line_items=line_items,
        total=total,
    )

File: backend/tools/budget.js
import { readJson, sendJson } from "../shared/http.js";
import { assertObject, optionalNumber, optionalStringArray } from "../shared/validation.js";

function estimateBudget(payload) {
  assertObject(payload);

  const guestCount = optionalNumber(payload.guestCount, 8, "guestCount");
  const perGuest = optionalNumber(payload.perGuest, 14, "perGuest");
  const categories = optionalStringArray(payload.categories, ["drinks", "snacks", "supplies"], "categories");

  const lineItems = categories.map((category) => {
    const multiplier = category === "decor" ? 0.25 : 1;
    const estimate = Math.round(guestCount * perGuest * multiplier * 100) / 100;

    return {
      category,
      estimate
    };
  });

  const total = lineItems.reduce((sum, item) => sum + item.estimate, 0);

  return {
    guestCount,
    perGuest,
    currency: "EUR",
    lineItems,
    total: Math.round(total * 100) / 100
  };
}

export const budgetTool = {
  id: "budget",
  name: "Budget",
  capabilities: [
    {
      id: "budget.estimate",
      tool: "budget",
      name: "Estimate Budget",
      description: "Estimate cost from guest count, budget target, and categories.",
      method: "POST",
      endpoint: "/budget/estimate",
      requestSchema: {
        type: "object",
        properties: {
          guestCount: { type: "number", minimum: 1 },
          perGuest: { type: "number", minimum: 0 },
          categories: {
            type: "array",
            items: { type: "string" }
          }
        }
      },
      responseSchema: {
        type: "object",
        properties: {
          guestCount: { type: "number" },
          perGuest: { type: "number" },
          currency: { type: "string" },
          lineItems: {
            type: "array",
            items: { type: "object" }
          },
          total: { type: "number" }
        },
        required: ["guestCount", "perGuest", "currency", "lineItems", "total"]
      }
    }
  ],
  async route({ request, response, url }) {
    if (request.method !== "POST" || url.pathname !== "/budget/estimate") {
      return false;
    }

    const payload = await readJson(request);
    sendJson(response, 200, estimateBudget(payload));
    return true;
  }
};
